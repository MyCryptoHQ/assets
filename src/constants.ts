import { number, object, string } from '@hapi/joi';
import { resolve } from 'path';

export interface Asset {
  name: string;
  symbol: string;
  uuid: string;
  address?: string;
  decimal?: string;
}

export interface ParsedAsset {
  coinCapId?: string;
  coinGeckoId?: string;
  cryptoCompareId?: string;
  cryptoCurrencyIconName?: string;
  dexAgId?: string;
}

/**
 * Hardcoded base assets. The UUIDs are generated from the chain ID, e.g. `uuid('1')`.
 */
export const BASE_ASSETS: Asset[] = [
  {
    name: 'Ethereum',
    symbol: 'ETH',
    uuid: '356a192b-7913-504c-9457-4d18c28d46e6'
  },
  {
    name: 'Ethereum Classic',
    symbol: 'ETC',
    uuid: '6c1e671f-9af5-546d-9c1a-52067bdf0e53'
  }
];

/**
 * Token asset validation schema.
 */
export const TOKEN_ASSET_SCHEMA = object({
  symbol: string().required(),
  address: string()
    .regex(/^0x[a-fA-F0-9]{40}$/)
    .required(),
  decimal: number()
    .min(0)
    .required(),
  name: string().required(),
  uuid: string().uuid()
});

/**
 * Parsed asset validation schema.
 */
export const ASSET_SCHEMA = object({
  coinCapId: string().optional(),
  coinGeckoId: string().optional(),
  cryptoCompareId: string().optional(),
  cryptoCurrencyIconName: string().optional(),
  dexAgId: string().optional()
});

export const OUTPUT_PATH = resolve(__dirname, '../assets/assets.json');
