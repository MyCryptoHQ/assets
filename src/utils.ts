import { compareTwoStrings } from 'string-similarity';
import { Asset, TOKEN_ASSET_SCHEMA } from './constants';

/**
 * Checks whether two strings are similar to each other, based on the Sørensen–Dice coefficient. Returns `true` if the
 * strings are similar, and `false` otherwise.
 *
 * @param {string} first
 * @param {string} second
 * @return {boolean}
 */
export const isSimilar = (first: string, second: string): boolean => {
  return compareTwoStrings(first, second) >= 0.5;
};

/**
 * Checks if the token is a valid token according to the validation schema.
 *
 * @param {Asset} token
 * @return {boolean}
 */
export const isValidToken = (token: Asset): boolean => {
  return !TOKEN_ASSET_SCHEMA.validate(token).error;
};

/**
 * Get the keys of an object as a strong typed array.
 *
 * @param {Object} object
 * @return {(keyof T)[]}
 * @template T
 */
export const keys = <T>(object: T): (keyof T)[] => {
  return Object.keys(object) as (keyof T)[];
};

/**
 * Checks if an object is empty.
 *
 * @param {object} object
 * @return {boolean}
 */
export const isEmpty = (object: object): boolean => {
  return Object.keys(object).length === 0;
};
