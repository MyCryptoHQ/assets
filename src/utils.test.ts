import { isEmpty, isSimilar, isValidAsset, isValidToken, keys } from './utils';
import { ParsedAsset } from './constants';

describe('isSimilar', () => {
  it('returns true for similar strings', () => {
    expect(isSimilar('Ethereum', 'Ethereum Network')).toBeTruthy();
    expect(isSimilar('Golem', 'Golem Token')).toBeTruthy();
  });

  it('returns false for non-similar strings', () => {
    expect(isSimilar('Ethereum', 'Bitcoin')).toBeFalsy();
    expect(isSimilar('Ethereum Network', 'Golem Network Token')).toBeFalsy();
  });
});

describe('isValildToken', () => {
  it('checks if a token is valid', () => {
    expect(
      isValidToken({
        symbol: 'FOO',
        address: '0x0000000000000000000000000000000000000000',
        decimal: '18',
        name: 'Foo',
        uuid: '6b4dbfa6-8ee7-452f-9ca9-ad9663f50c62'
      })
    ).toBeTruthy();
  });

  it('checks if a token is invalid', () => {
    expect(isValidToken({} as any)).toBeFalsy();

    expect(
      isValidToken({
        symbol: 'FOO',
        decimal: '18',
        name: 'Foo',
        uuid: '6b4dbfa6-8ee7-452f-9ca9-ad9663f50c62'
      })
    ).toBeFalsy();

    expect(
      isValidToken({
        symbol: 'FOO',
        address: '0x0',
        decimal: '18',
        name: 'Foo',
        uuid: '6b4dbfa6-8ee7-452f-9ca9-ad9663f50c62'
      })
    ).toBeFalsy();

    expect(
      isValidToken({
        symbol: 'FOO',
        address: '0x0000000000000000000000000000000000000000',
        decimal: '18',
        name: 'Foo',
        uuid: 'foo-bar'
      })
    ).toBeFalsy();

    expect(
      isValidToken({
        symbol: 'FOO',
        address: '0x0000000000000000000000000000000000000000',
        decimal: '-1',
        name: 'Foo',
        uuid: '6b4dbfa6-8ee7-452f-9ca9-ad9663f50c62'
      })
    ).toBeFalsy();
  });
});

describe('isValidAsset', () => {
  it('checks if an asset is valid', () => {
    expect(
      isValidAsset({
        coinCapId: 'foo',
        coinGeckoId: 'bar',
        cryptoCompareId: 'baz',
        cryptoCurrencyIconName: 'qux'
      })
    ).toBeTruthy();

    expect(isValidAsset({})).toBeTruthy();
  });

  it('checks if an asset is invalid', () => {
    expect(
      isValidAsset({
        coinCapId: 'foo',
        coinGeckoId: 'bar',
        cryptoCompareId: 'baz',
        cryptoCurrencyIconName: 'qux',
        invalid: 'field'
      } as any)
    ).toBeFalsy();
  });
});

describe('keys', () => {
  it('returns the key of an object', () => {
    const object: ParsedAsset = {
      coinCapId: 'foo',
      coinGeckoId: 'bar',
      cryptoCompareId: 'baz',
      cryptoCurrencyIconName: 'qux'
    };

    const objectKeys = keys(object);
    expect(Object.keys(object)).toEqual(objectKeys);
    expect(objectKeys).toMatchSnapshot();
  });

  it('has proper TypeScript types', () => {
    const object: ParsedAsset = {
      coinCapId: 'foo',
      coinGeckoId: 'bar',
      cryptoCompareId: 'baz',
      cryptoCurrencyIconName: 'qux'
    };

    const objectKeys = keys(object);
    expect(object[objectKeys[0]]).toBeTruthy();
    expect(object[objectKeys[1]]).toBeTruthy();
    expect(object[objectKeys[2]]).toBeTruthy();
    expect(object[objectKeys[3]]).toBeTruthy();
  });
});

describe('isEmpty', () => {
  it('checks if an object is empty', () => {
    const object: ParsedAsset = {
      coinCapId: 'foo',
      coinGeckoId: 'bar',
      cryptoCompareId: 'baz',
      cryptoCurrencyIconName: 'qux'
    };

    expect(isEmpty(object)).toBeFalsy();
    expect(isEmpty({})).toBeTruthy();
  });
});
