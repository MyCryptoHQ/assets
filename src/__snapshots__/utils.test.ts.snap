// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`keys returns the key of an object 1`] = `
Array [
  "coinCapId",
  "coinGeckoId",
  "cryptoCompareId",
  "cryptoCurrencyIconName",
]
`;
